crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -460,17 +463,184 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
+    upgrade_ticket_arg: Argument,
 ) -> Result<Vec<Value>, ExecutionError> {
     // Check that package upgrades are supported.
     context
         .protocol_config
         .check_package_upgrades_supported()
         .map_err(|_| ExecutionError::from_kind(ExecutionErrorKind::FeatureNotYetSupported))?;
 
-    invariant_violation!("Package upgrades are turned off. We should NEVER get here.")
+    assert_invariant!(
+        !module_bytes.is_empty(),
+        "empty package is checked in transaction input checker"
+    );
+
+    let upgrade_ticket: UpgradeTicket = {
+        let mut ticket_bytes = Vec::new();
+        let ticket_val: Value =
+            context.by_value_arg(CommandKind::Upgrade, 0, upgrade_ticket_arg)?;
+        let ticket_type = context
+            .session
+            .load_type(&TypeTag::Struct(Box::new(UpgradeTicket::type_())))
+            .map_err(|e| context.convert_vm_error(e))?;
+        check_param_type::<_, _, Mode>(context, 0, &ticket_val, &ticket_type)?;
+        ticket_val.write_bcs_bytes(&mut ticket_bytes);
+        bcs::from_bytes(&ticket_bytes).map_err(|_| {
+            ExecutionError::from_kind(ExecutionErrorKind::CommandArgumentError {
+                arg_idx: 0,
+                kind: CommandArgumentError::InvalidBCSBytes,
+            })
+        })?
+    };
+
+    // Check digest.
+    let computed_digest =
+        MovePackage::compute_digest_for_modules_and_deps(&module_bytes, &dep_ids).to_vec();
+    if computed_digest != upgrade_ticket.digest {
+        return Err(ExecutionError::from_kind(
+            ExecutionErrorKind::PackageUpgradeError {
+                upgrade_error: PackageUpgradeError::DigestDoesNotMatch {
+                    digest: computed_digest,
+                },
+            },
+        ));
+    }
+
+    // Run the move + sui verifier on the modules and publish them into the cache.
+    // This will give use the new package ID as well that we will need when creating
+    // the linkage tables when upgrading the package.
+    let (upgraded_package_modules, new_package_id) =
+        publish_and_verify_modules::<_, _, Mode>(context, module_bytes)?;
+
+    // Check that this package ID points to a package and get the package we're upgrading.
+    let current_package = fetch_package(context, &upgrade_ticket.package.bytes)?;
+
+    // Full backwards compatibility except that we allow friend function signatures to change.
+    let compatiblity_checker = {
+        let check_struct_and_pub_function_linking = true;
+        let check_struct_layout = true;
+        let check_friend_linking = false;
+        Compatibility::new(
+            check_struct_and_pub_function_linking,
+            check_struct_layout,
+            check_friend_linking,
+        )
+    };
+    check_compatibility(
+        &current_package,
+        &upgraded_package_modules,
+        compatiblity_checker,
+        upgrade_ticket.policy,
+    )?;
+
+    // Deduplicate any dependency ids.
+    dep_ids.dedup();
```

Comment:
Where does this list of ids come from? Is it user input? We could just check for uniqueness in the transaction input checker?

Reply:
I don't think this does what you want in any case -- it only deduplicates adjacent elements:

> Removes consecutive repeated elements in the vector according to the [PartialEq](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html) trait implementation.

Better to do something like:

```suggestion
    let dep_ids: BTreeSet<_> = dep_ids.into_iter().collect();
```

@tnowacki, I don't think there's much of a need to check this during signing -- if we're doing the work, we might as well get paid gas for it.